LanguageParser.hs

```diff
@@ -1,40 +1,414 @@
+listOfSentenceP = some sentenceP
+
+ifSoP :: Parser Sentence
+ifSoP = IfSo <$> (P.string' "If" *> P.space1 *> P.string' "so" *>
+                 P.space *> P.char ',' *> P.space *> sentenceP')
+
+ifNotP :: Parser Sentence
+ifNotP = IfNot <$> (P.string' "If" *> P.space1 *> P.string' "not" *>
+                   P.space *> P.char ',' *> P.space *> sentenceP')
+
+outputNumberP :: Parser Sentence
+outputNumberP = constP OutputNumber
+                (P.string' "Open" <* P.space1 <* oneOfSecondPersonPos <*
+                P.space1 <* P.string' "heart")
+
+outputCharacterP :: Parser Sentence
+outputCharacterP = constP OutputCharacter
+                   (P.string' "Speak" <* P.space1 <* oneOfSecondPersonPos <*
+                   P.space1 <* P.string' "mind")
+
+inputNumberP :: Parser Sentence
+inputNumberP = constP InputNumber
+               (P.string' "Listen" <* P.space1 <* P.string "to" <* P.space1 <*
+               oneOfSecondPersonPos <* P.space1 <* P.string' "heart")
+
+inputCharacterP :: Parser Sentence
+inputCharacterP = constP InputCharacter
+                  (P.string' "Open" <* P.space1 <* oneOfSecondPersonPos <*
+                  P.space1 <* P.string' "mind")
+
+declarationP :: Parser Sentence
+declarationP =
+  P.try decVarient1 <|> P.try decVarient2 <|> decVarient3
+  where
+
+    decVarient1 :: Parser Sentence
+    decVarient1 = Declaration <$> (oneOfString' W.secondPerson *> P.space1 *>
+                                  oneOfString' W.be *> P.space1 *>
+                                  P.string' "as" *> P.space1 *>
+                                  oneOfString' W.adjectives *> P.space1 *>
+                                  P.string' "as" *> P.space1 *> expressionP)
+
+    decVarient2 :: Parser Sentence
+    decVarient2 = Declaration <$> (oneOfString' W.secondPerson *> P.space1 *>
+                                  oneOfString' W.be *> P.space1 *> expressionP)
+
+    decVarient3 :: Parser Sentence
+    decVarient3 = Declaration <$> (oneOfString' W.secondPerson *> P.space1 *>
+                                  expressionP)
+
+pushP :: Parser Sentence
+pushP = Push <$> (P.string' "Remember" *> P.space1 *> referenceP)
+
+popP :: Parser Sentence
+popP = constP Pop (P.string' "Recall") <* parseUntilEndPunc
+
+genericGoTo :: (Label -> Sentence) -> String -> Parser Sentence
+genericGoTo con s = con <$> (oneOfString' ["Let us", "We shall", "We must"] *>
+                    P.space1 *> oneOfString' ["return to", "proceed to"] *>
+                    P.space1 *> P.string' s *> P.space1 *>
+                    labelP parseUntilEndPunc)
+
+goToSceneP :: Parser Sentence
+goToSceneP = genericGoTo GotoScene "scene"
+
+goToActP :: Parser Sentence
+goToActP = genericGoTo GotoAct "act"
+
+-- TODO: this does NOT enforce an ending of ? right now. I think this is fine, let me know what you think.
+conditionalP :: Parser Sentence
+conditionalP = Conditional <$> comparisonP
+
+sentenceP' :: Parser Sentence
+sentenceP' = P.try ifSoP <|>
```

Comment:
Nitpicky, but could be aesthetically nice to line up the alternates

Reply:
auto-formatter

also what do you mean

Reply:
```haskell
             P.try ifNotP                    <|>
             P.try outputNumberP    <|>
             P.try outputCharacterP <|>
             P.try inputNumberP       <|>
             P.try inputCharacterP    <|>
             P.try declarationP          <|>
```